File: iterableapi-ui/src/main/java/com/iterable/iterableapi/ui/inbox/IterableInboxMessageFragment.java
package com.iterable.iterableapi.ui.inbox;

import android.net.Uri;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.fragment.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.iterable.iterableapi.IterableApi;
import com.iterable.iterableapi.IterableInAppLocation;
import com.iterable.iterableapi.IterableInAppMessage;
import com.iterable.iterableapi.IterableUtil;
import com.iterable.iterableapi.ui.R;

import java.util.List;

public class IterableInboxMessageFragment extends Fragment {
    public static final String ARG_MESSAGE_ID = "messageId";
    public static final String STATE_LOADED = "loaded";

    private String messageId;
    private WebView webView;
    private IterableInAppMessage message;
    private boolean loaded = false;

    public static IterableInboxMessageFragment newInstance(String messageId) {
        IterableInboxMessageFragment fragment = new IterableInboxMessageFragment();

        Bundle args = new Bundle();
        args.putString(ARG_MESSAGE_ID, messageId);
        fragment.setArguments(args);

        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            messageId = getArguments().getString(ARG_MESSAGE_ID);
        }
        if (savedInstanceState != null) {
            loaded = savedInstanceState.getBoolean(STATE_LOADED, false);
        }
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBoolean(STATE_LOADED, true);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view  = inflater.inflate(R.layout.iterable_inbox_message_fragment, container, false);
        webView = view.findViewById(R.id.webView);
        loadMessage();
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        ViewCompat.setOnApplyWindowInsetsListener(view, (v, insets) -> {
            Insets bars = insets.getInsets(WindowInsetsCompat.Type.systemBars() | WindowInsetsCompat.Type.displayCutout());
            v.setPadding(0, bars.top, 0, bars.bottom);
            return insets;
        });

        ViewCompat.requestApplyInsets(view);
    }


    private IterableInAppMessage getMessageById(String messageId) {
        List<IterableInAppMessage> messages = IterableApi.getInstance().getInAppManager().getMessages();
        for (IterableInAppMessage message : messages) {
            if (message.getMessageId().equals(messageId)) {
                return message;
            }
        }
        return null;
    }

    private void loadMessage() {
        message = getMessageById(messageId);
        if (message != null) {
            // Use configured base URL to enable CORS for external resources (e.g., custom fonts)
            webView.loadDataWithBaseURL(IterableUtil.getWebViewBaseUrl(), message.getContent().html, "text/html", "UTF-8", "");
            webView.setWebViewClient(webViewClient);
            if (!loaded) {
                IterableApi.getInstance().trackInAppOpen(message, IterableInAppLocation.INBOX);
                loaded = true;
            }
            if (getActivity() != null) {
                getActivity().setTitle(message.getInboxMetadata().title);
            }
        }
    }

    private WebViewClient webViewClient = new WebViewClient() {
        @Override
        public boolean shouldOverrideUrlLoading(WebView view, String url) {
            IterableApi.getInstance().trackInAppClick(message, url, IterableInAppLocation.INBOX);
            IterableApi.getInstance().getInAppManager().handleInAppClick(message, Uri.parse(url));
            if (getActivity() != null) {
                getActivity().finish();
            }
            return true;
        }
    };
}

File: iterableapi-ui/src/main/java/com/iterable/iterableapi/ui/inbox/IterableInboxMessageFragment.kt
package com.iterable.iterableapi.ui.inbox

import android.net.Uri
import android.os.Bundle
import androidx.annotation.NonNull
import androidx.annotation.Nullable
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.WebView
import android.webkit.WebViewClient

import com.iterable.iterableapi.IterableApi
import com.iterable.iterableapi.IterableInAppLocation
import com.iterable.iterableapi.IterableInAppMessage
import com.iterable.iterableapi.ui.R

class IterableInboxMessageFragment : Fragment() {
    
    companion object {
        const val ARG_MESSAGE_ID = "messageId"
        const val STATE_LOADED = "loaded"
        
        @JvmStatic
        fun newInstance(messageId: String?): IterableInboxMessageFragment {
            val fragment = IterableInboxMessageFragment()
            val args = Bundle()
            args.putString(ARG_MESSAGE_ID, messageId)
            fragment.arguments = args
            return fragment
        }
    }

    private var messageId: String? = null
    private lateinit var webView: WebView
    private var message: IterableInAppMessage? = null
    private var loaded = false

    override fun onCreate(@Nullable savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.let {
            messageId = it.getString(ARG_MESSAGE_ID)
        }
        savedInstanceState?.let {
            loaded = it.getBoolean(STATE_LOADED, false)
        }
    }

    override fun onSaveInstanceState(@NonNull outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBoolean(STATE_LOADED, true)
    }

    @Nullable
    override fun onCreateView(@NonNull inflater: LayoutInflater, @Nullable container: ViewGroup?, @Nullable savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.iterable_inbox_message_fragment, container, false)
        webView = view.findViewById(R.id.webView)
        loadMessage()
        return view
    }

    private fun getMessageById(messageId: String?): IterableInAppMessage? {
        val messages = IterableApi.getInstance().inAppManager.messages
        for (message in messages) {
            if (message.messageId == messageId) {
                return message
            }
        }
        return null
    }

    private fun loadMessage() {
        message = getMessageById(messageId)
        message?.let { msg ->
            webView.loadDataWithBaseURL("", msg.content.html, "text/html", "UTF-8", "")
            webView.webViewClient = webViewClient
            if (!loaded) {
                IterableApi.getInstance().trackInAppOpen(msg, IterableInAppLocation.INBOX)
                loaded = true
            }
            activity?.setTitle(msg.inboxMetadata.title)
        }
    }

    private val webViewClient = object : WebViewClient() {
        override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
            message?.let { msg ->
                IterableApi.getInstance().trackInAppClick(msg, url, IterableInAppLocation.INBOX)
                IterableApi.getInstance().inAppManager.handleInAppClick(msg, Uri.parse(url))
                activity?.finish()
            }
            return true
        }
    }
}
